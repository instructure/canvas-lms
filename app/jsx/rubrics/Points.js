/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import _ from 'lodash'
import PropTypes from 'prop-types'
import ScreenReaderContent from '@instructure/ui-a11y/lib/components/ScreenReaderContent'
import TextInput from '@instructure/ui-forms/lib/components/TextInput'
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex'
import I18n from 'i18n!edit_rubric'

import { assessmentShape } from './types'

export const roundIfWhole = (n) => (
  I18n.toNumber(n, { precision: Math.floor(n) === n ? 0 : 1 })
)
const pointString = (n) => n !== undefined ? roundIfWhole(n) : '--'

export const possibleString = (possible) =>
  I18n.t('%{possible} pts', {
    possible: I18n.toNumber(possible, { precision : 1 })
  })

export const scoreString = (points, possible) =>
  I18n.t('%{points} / %{possible}', {
    points: pointString(points),
    possible: possibleString(possible)
  })

const invalid = () => [{ text: I18n.t('Invalid score'), type: 'error' }]
const pointError = (points, pointsText) =>
  (_.isNil(points) && pointsText) ? invalid() : []

const noExtraCredit = () => [
  { text: I18n.t('Cannot give outcomes extra credit'), type: 'error' }
]
const extraCreditError = (points, possible, allowExtraCredit) =>
  !allowExtraCredit && (points > possible) ? noExtraCredit() : []

const Points = (props) => {
  const {
    allowExtraCredit,
    assessing,
    assessment,
    onPointChange,
    pointsPossible
  } = props

  if (assessment === null) {
    return (
      <div className="react-rubric-cell graded-points">
        {possibleString(pointsPossible)}
      </div>
    )
  } else {
    const points = _.get(assessment, 'points')
    const pointsText = _.get(assessment, 'pointsText')
    if (!assessing) {
      return (
        <div className="react-rubric-cell graded-points">
          {scoreString(points, pointsPossible)}
        </div>
      )
    } else {
      const usePointsText = pointsText !== null && pointsText !== undefined
      return (
        <div className="react-rubric-cell graded-points">
          <Flex alignItems="start">
            <FlexItem size="4rem">
              <TextInput
                inline
                label={<ScreenReaderContent>{I18n.t('Points')}</ScreenReaderContent>}
                messages={[
                  ...pointError(points, pointsText),
                  ...extraCreditError(points, pointsPossible, allowExtraCredit)
                ]}
                onChange={(e) => onPointChange(e.target.value)}
                value={usePointsText ? pointsText : pointString(points)}
                width="4rem"
              />
            </FlexItem>
            <FlexItem margin="small">
              {`/ ${possibleString(pointsPossible)}`}
            </FlexItem>
          </Flex>
        </div>
      )
    }
  }
}
Points.propTypes = {
  allowExtraCredit: PropTypes.bool,
  assessing: PropTypes.bool,
  assessment: PropTypes.shape(assessmentShape),
  onPointChange: PropTypes.func,
  pointsPossible: PropTypes.number.isRequired,
}
Points.defaultProps = {
  allowExtraCredit: true,
  assessing: false,
  onPointChange: null
}

export default Points
