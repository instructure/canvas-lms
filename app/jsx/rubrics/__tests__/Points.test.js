/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import { shallow } from 'enzyme'
import Points from '../Points'

describe('The Points component', () => {
  const component = (props) => shallow(<Points {...props} />)
  const id = { criterion_id: '_7506' }

  it('renders the root component as expected', () => {
    expect(component({
      assessment: { ...id, points: 1, pointsText: '1' },
      pointsPossible: 2
    }).debug()).toMatchSnapshot()
  })

  it('renders as expected with fractional points', () => {
    expect(component({
      assessment: { ...id, points: 1.1, pointsText: '1.1' },
      pointsPossible: 2
    }).debug()).toMatchSnapshot()
  })

  it('renders blank when points are undefined', () => {
    expect(component({
      assessing: true,
      assessment: { ...id, pointsText: '' },
      pointsPossible: 2
    }).debug()).toMatchSnapshot()
  })

  it('renders points possible with no assessment', () => {
    expect(component({
      assessing: false,
      assessment: null,
      pointsPossible: 2
    }).debug()).toMatchSnapshot()
  })

  const withText = (pointsText, points) => component({
    allowExtraCredit: false,
    assessing: true,
    assessment: {
      ...id,
      points,
      pointsText,
    },
    pointsPossible: 5
  })

  it('renders an error when points is a string', () => {
    const el = withText('stringy')
    expect(el.debug()).toMatchSnapshot()
    expect(el.find('TextInput').prop('messages')).toHaveLength(1)
  })

  it('renders an error when extra credit cannot be given', () => {
    const el = withText('30', 30)
    expect(el.debug()).toMatchSnapshot()
    expect(el.find('TextInput').prop('messages')).toHaveLength(1)
  })

  it('renders no error with a blank string', () => {
    const expectNoErrorsWith = (t, p) =>
      expect(withText(t, p).find('TextInput').prop('messages')).toHaveLength(0)

    expectNoErrorsWith('')
    expectNoErrorsWith(null)
    expectNoErrorsWith(undefined)
    expectNoErrorsWith('0', 0)
    expectNoErrorsWith('2.2', 2.2)
  })
})
