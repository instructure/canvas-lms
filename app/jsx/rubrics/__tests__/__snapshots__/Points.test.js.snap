// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Points component renders an error when extra credit cannot be given 1`] = `
"<div className=\\"react-rubric-cell graded-points\\">
  <Flex alignItems=\\"start\\" as=\\"span\\" direction=\\"row\\" justifyItems=\\"start\\" inline={false} visualDebug={false} wrapItems={false}>
    <FlexItem size=\\"4rem\\" as=\\"span\\" grow={false} shrink={false}>
      <TextInput inline={true} label={{...}} messages={{...}} onChange={[Function]} value=\\"30\\" width=\\"4rem\\" type=\\"text\\" size=\\"medium\\" textAlign=\\"start\\" disabled={false} readOnly={false} inputRef={[Function]} layout=\\"stacked\\" />
    </FlexItem>
    <FlexItem margin=\\"small\\" as=\\"span\\" grow={false} shrink={false}>
      / 5.0 pts
    </FlexItem>
  </Flex>
</div>"
`;

exports[`The Points component renders an error when points is a string 1`] = `
"<div className=\\"react-rubric-cell graded-points\\">
  <Flex alignItems=\\"start\\" as=\\"span\\" direction=\\"row\\" justifyItems=\\"start\\" inline={false} visualDebug={false} wrapItems={false}>
    <FlexItem size=\\"4rem\\" as=\\"span\\" grow={false} shrink={false}>
      <TextInput inline={true} label={{...}} messages={{...}} onChange={[Function]} value=\\"stringy\\" width=\\"4rem\\" type=\\"text\\" size=\\"medium\\" textAlign=\\"start\\" disabled={false} readOnly={false} inputRef={[Function]} layout=\\"stacked\\" />
    </FlexItem>
    <FlexItem margin=\\"small\\" as=\\"span\\" grow={false} shrink={false}>
      / 5.0 pts
    </FlexItem>
  </Flex>
</div>"
`;

exports[`The Points component renders as expected with fractional points 1`] = `
"<div className=\\"react-rubric-cell graded-points\\">
  1.1 / 2.0 pts
</div>"
`;

exports[`The Points component renders blank when points are undefined 1`] = `
"<div className=\\"react-rubric-cell graded-points\\">
  <Flex alignItems=\\"start\\" as=\\"span\\" direction=\\"row\\" justifyItems=\\"start\\" inline={false} visualDebug={false} wrapItems={false}>
    <FlexItem size=\\"4rem\\" as=\\"span\\" grow={false} shrink={false}>
      <TextInput inline={true} label={{...}} messages={{...}} onChange={[Function]} value=\\"\\" width=\\"4rem\\" type=\\"text\\" size=\\"medium\\" textAlign=\\"start\\" disabled={false} readOnly={false} inputRef={[Function]} layout=\\"stacked\\" />
    </FlexItem>
    <FlexItem margin=\\"small\\" as=\\"span\\" grow={false} shrink={false}>
      / 2.0 pts
    </FlexItem>
  </Flex>
</div>"
`;

exports[`The Points component renders points possible with no assessment 1`] = `
"<div className=\\"react-rubric-cell graded-points\\">
  2.0 pts
</div>"
`;

exports[`The Points component renders the root component as expected 1`] = `
"<div className=\\"react-rubric-cell graded-points\\">
  1 / 2.0 pts
</div>"
`;
